clash-cores/src/Clash/Cores/Floating/Xilinx/BlackBoxes.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RecordWildCards #-}
+module Clash.Cores.Floating.Xilinx.BlackBoxes
+  ( addFloatTclTF
+  ) where
+
+import Prelude
+
+import Control.Monad.State (State())
+import Data.Maybe (isJust, fromJust)
+import Data.String (fromString)
+import Data.String.Interpolate (i)
+
+import Clash.Backend (Backend())
+import Clash.Netlist.Types
+  (BlackBoxContext(..), Expr(..), HWType(..), Literal(..), Modifier(..),
+   TemplateFunction(..))
+import Data.Text.Prettyprint.Doc.Extra (Doc)
+
+data HasCustom = HasCustom
+  { addSubVal ::  !(Maybe String)
+  , hasArchOpt :: !Bool
+  , hasDspUsage :: !Bool
+  , hasBMemUsage :: !Bool
+  }
+
+defHasCustom :: HasCustom
+defHasCustom = HasCustom
+  { addSubVal = Nothing
+  , hasArchOpt = True
+  , hasDspUsage = True
+  , hasBMemUsage = False
+  }
+
+addFloatTclTF :: TemplateFunction
+addFloatTclTF =
+  binaryFloatTclTF (defHasCustom { addSubVal = Just "Add" }) "Add_Subtract"
+
+binaryFloatTclTF
+  :: HasCustom
+  -> String
+  -> TemplateFunction
+binaryFloatTclTF hasCustom operType =
+  TemplateFunction used valid (floatTclTemplate hasCustom operType)
+ where
+  used = [0..4]
+  valid = const True
+
+floatTclTemplate
+  :: Backend s
+  => HasCustom
+  -> String
+  -> BlackBoxContext
+  -> State s Doc
+floatTclTemplate (HasCustom {..}) operType bbCtx = pure bbText
+ where
+  compName = bbQsysIncName bbCtx !! 0
+
+  (Literal _ (NumLit latency), _, _) = bbInputs bbCtx !! 1
+  (DataCon _ _ cfgExprs, _, _) = bbInputs bbCtx !! 3
+  cfgArchOptExpr = cfgExprs !! 0
+  cfgDspUsageExpr = cfgExprs !! 1
+  cfgBMemUsageExpr = cfgExprs !! 2
+
+  DataCon _ (DC (Sum _ cfgArchOptConstrs, cfgArchOptTag)) _ = cfgArchOptExpr
+  cfgArchOpt = cfgArchOptConstrs !! cfgArchOptTag
+  tclArchOpt :: String
+  tclArchOpt =
+    case cfgArchOpt of
+      "Clash.Cores.Floating.Xilinx.Internal.SpeedArch" -> "Speed_Optimized"
+      "Clash.Cores.Floating.Xilinx.Internal.LatencyArch" -> "Low_Latency"
+      _ -> error "Unknown FloatingArchOpt constructor"
+
+  DataCon _ (DC (Sum _ cfgDspUsageConstrs, cfgDspUsageTag)) _ = cfgDspUsageExpr
+  cfgDspUsage = cfgDspUsageConstrs !! cfgDspUsageTag
+  tclDspUsage :: String
+  tclDspUsage =
+    case cfgDspUsage of
+      "Clash.Cores.Floating.Xilinx.Internal.NoDspUsage" -> "No_Usage"
+      "Clash.Cores.Floating.Xilinx.Internal.MediumDspUsage" -> "Medium_Usage"
+      "Clash.Cores.Floating.Xilinx.Internal.FullDspUsage" -> "Full_Usage"
+      "Clash.Cores.Floating.Xilinx.Internal.MaxDspUsage" -> "Max_Usage"
+      _ -> error "Unknown FloatingDspUsage constructor"
```

Comment:
To make refactors easier I'd suggest:

```haskell

  tclDspUsage
    | cfgDspUsage == show 'NoDspUsage -> "No_Usage"
    | cfgDspUsage == show 'MediumDspUsage -> "Medium_Usage"
    ...
    | otherwise -> error ...
```

What do you think?

Reply:
Looks good! Thanks for the hint!